### PR TITLE
Fix restart crash

### PR DESCRIPTION
**Pull request by roberthoffmann**
Do not close inotify desciptors when restarting the server. This
may lead to crashes in glib.

Fixes crash in libglib/libgio with GLib-GIO-ERROR : inotify read(): Bad file descriptor

See also src/tests/ibus-bus.c: start_exit_async():
    /* When `./runtest ibus-bus` runs, ibus-daemon sometimes failed to
     * restart because closing a file descriptor was failed in
     * bus/server.c:_restart_server() with a following error:
     *     "inotify read(): Bad file descriptor"

Test plan:
Run
$ ibus restart
several times. Without the patch ibus-daemon will crash occasionally.

### CODE REVIEW

File: bus/server.c [hunk 1]
```diff
@@ -26,6 +26,9 @@
 #include <glib/gstdio.h>
 #include <gio/gio.h>
 #include <stdlib.h>
+#include <stdio.h>
```
[the remark this answers is not in the document]
Comment (roberthoffmann) [reply bus/server.c]: Possibly

File: bus/server.c [hunk 2]
```diff
@@ -45,6 +48,10 @@ _restart_server (void)
 {
     gchar *exe;
     gint fd;
+    ssize_t r;
+    int MAXSIZE = 0xFFF;
+    char proclnk[MAXSIZE];
+    char filename[MAXSIZE];
 
     exe = g_strdup_printf ("/proc/%d/exe", getpid ());
     exe = g_file_read_link (exe, NULL);
```
[the remark this answers is not in the document]
Comment (roberthoffmann) [reply bus/server.c]: Yes, changed.

File: bus/server.c [hunk 3]
```diff
@@ -54,7 +61,21 @@ _restart_server (void)
```
[the remark this answers is not in the document]
Comment (roberthoffmann) [reply bus/server.c]: changed

File: bus/server.c [hunk 4]
```diff
@@ -54,7 +61,21 @@ _restart_server (void)
 
     /* close all fds except stdin, stdout, stderr */
     for (fd = 3; fd <= sysconf (_SC_OPEN_MAX); fd ++) {
-        close (fd);
+        errno = 0;
+        /* only close valid fds */
+        if (fcntl(fd, F_GETFD) != -1 || errno != EBADF) {
```
[the remark this answers is not in the document]
Comment (roberthoffmann) [reply bus/server.c]: done

File: bus/server.c [hunk 5]
```diff
@@ -54,7 +61,21 @@ _restart_server (void)
...
+            sprintf(proclnk, "/proc/self/fd/%d", fd);
```
[the remark this answers is not in the document]
Comment (roberthoffmann) [reply bus/server.c]: done

File: bus/server.c [hunk 6]
```diff
@@ -54,7 +61,21 @@ _restart_server (void)
...
+            r = readlink(proclnk, filename, MAXSIZE);
```
[the remark this answers is not in the document]
Comment (roberthoffmann) [reply bus/server.c]: done

File: bus/server.c [hunk 7]
```diff
@@ -54,7 +61,21 @@ _restart_server (void)
...
+            if (r < 0) {
+                continue;
+            }
+            filename[r] = '\0';
+
+            /* Do not close 'anon_inode:inotify' fds, that may crash in glib */
+            if (strcmp(filename, "anon_inode:inotify") != 0) {
```
[the remark this answers is not in the document]
Comment (roberthoffmann) [reply bus/server.c]: done

**Comment by fujiwarat** [comment]
Thank you for the patch.

**Pull request closed without merging**